File: 0740-delete-and-earn/0740-delete-and-earn.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        m = max(nums)
        v = [0] * (m + 1)
        hi = max(nums)
        for i in nums:
            v[i] += i
        a, b = 0, v[1]
        for i in range(2, hi + 1):
            a, b = b, max(a + v[i], b)
        return b

File: 0740-delete-and-earn/0740-delete-and-earn.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int n = nums.size();
        vector<int> v(1e4+1);
        for (int i=0; i<n; i++) v[nums[i]]++;
        vector<int> dp(1e4+1, -1);
        function<int(int)> dfs = [&](int i) {
            if (i < 0 or i > 1e4) return 0;
            if (dp[i] != -1) return dp[i];
            dp[i] = max(dfs(i+1), i * v[i] + dfs(i+2));
            return dp[i];
        };
        dfs(0);
        // for (int i=0; i<5; i++) cout << v[i] << ' ';
        // cout << endl;
        // for (int i=0; i<5; i++) cout << dp[i] << ' ';
        // cout << endl;
        return dp[0];
    }
};